core/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java

```diff
@@ -234,7 +233,8 @@ String checkWriteConsistency() {
     }
 
     private void decPendingAndFinishIfNeeded() {
-        if (pendingShards.decrementAndGet() <= 0) {
+        assert pendingShards.get() > 0;
+        if (pendingShards.decrementAndGet() == 0) {
```

Comment:
nit: I have nothing against being defensive here with `pendingShards.decrementAndGet() <= 0` but I would like the assertion to capture `== 0` check. Preferably on the same variable as pendingShards is modified concurrently. This means:

```
int pendingShardsCount = pendingShards.decrementAndGet()
if (pendingShardsCount <= 0) {
   assert pendingShardsCount == 0;
    ...
}
```